### PR TITLE
add Const data structure

### PR DESCRIPTION
**Pull request by julien-truffaut**

### CODE REVIEW

**Comment by julien-truffaut** [comment]
sure, also do you have an example of law checking tests?

**Comment by non** [comment]
I plan to use a strategy similar to the one we're using in algebra:

https://github.com/non/algebra/tree/master/laws/src/main/scala/algebra/laws

I need to get some basic examples included to sort of show how it will work; but feel free to push forward yourself if you want.

**Comment by non** [comment]
:+1:

**Comment by julien-truffaut** [comment]
I will add tests once we have them setup.

**Comment by julien-truffaut** [comment]
should I add `compare` and `partialCompare` method on `Const`?

**Comment by non** [comment]
Yes, I think those would both be good.

Also, I know it might get removed, but maybe you should extend `Bimonad` for now, since we have it?

**Comment by julien-truffaut** [comment]
@non `Const` is not a `Monad` or `Comonad`

File: data/src/main/scala/cats/data/Const.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      fa.traverse(f)
+
+    def foldLeft[A, B](fa: Const[C, A], b: B)(f: (B, A) => B): B = b
+
+    def foldRight[A, B](fa: Const[C, A], b: B)(f: (A, B) => B): B = b
+
+    def foldRight[A, B](fa: Const[C, A], b: Lazy[B])(f: (A, Lazy[B]) => B): Lazy[B] = b
+  }
+
+  implicit def constMoinoid[A: Monoid, B]: Monoid[Const[A, B]] = new Monoid[Const[A, B]]{
```
Comment (ceedubs) [inline data/src/main/scala/cats/data/Const.scala]: s/Moinoid/Monoid :)

Comment (julien-truffaut) [inline data/src/main/scala/cats/data/Const.scala]: how do you manage to see all my typo!

**Comment by non** [comment]
Oh huh, I think I must have gotten these issues confused, sorry! I think I was thinking of the `Function0` PR. My mistake.

**Comment by julien-truffaut** [comment]
np, basically I think it this PR is ready. Some tests are still missing but we can implement them once we have the other laws in place. What do you think?

**Comment by ceedubs** [comment]
I think it looks good and could probably be merged. Having said that, I'd really like to see an example for this. I've never actually used a Const structure, and seeing an example might help me understand how it could be useful.

**Comment by julien-truffaut** [comment]
a good example of `Const` is how to implement `foldMap` in terms of `traverse`:
```scala
def foldMap[F[_], A, B: Monoid](fa: F[A])(f: A => B): B = traverse(fa)(a => Const(f(a))).getConst
```
If you find it useful, where would you want me to add it?

**Comment by non** [comment]
I'm happy with this PR as it is. :+1:

I'd say we should merge it and then worry about the `foldMap` question in another PR.

**Comment by ceedubs** [comment]
@julien-truffaut ah that's a great example of Const. Thanks! Yeah that would be great to add somewhere, but where is a good question. It's concise enough I could see it going into the ScalaDoc for Const. I'm open to other ideas, though.

**Pull request merged**